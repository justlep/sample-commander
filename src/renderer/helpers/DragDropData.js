const SYM = Symbol();

/**
 * Represents a container for dragged fileItems and a drop-target dirItem,
 * i.e. everything required for some copy/move dialog.
 */
export default class DragDropData {
    /**
     * @param {FileItem[]} droppedFileItems
     * @param {DirItem} targetDirItem
     */
    constructor(droppedFileItems, targetDirItem) {
        this[SYM] = {
            fileItems: droppedFileItems || [],
            targetDirItem: targetDirItem
        };
    }
    /** @returns {FileItem[]} */
    getFileItems() {
        return this[SYM].fileItems;
    }
    /** @returns {DirItem} */
    getTargetDirItem() {
        return this[SYM].targetDirItem;
    }
}
