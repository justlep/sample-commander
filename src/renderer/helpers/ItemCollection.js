const SYM = Symbol();

/**
 * A simple wrapper for any array of objects, to be used in components data or store state while
 * preventing Vue from making the items themselves reactive.
 */
export default class ItemCollection {
    static from(items) {
        return new ItemCollection(items);
    }
    
    constructor(items) {
        this[SYM] = items || [];
    }
    getItems() {
        return this[SYM];
    }
}
