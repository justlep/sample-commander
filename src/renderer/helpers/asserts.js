
const {format} = require('util');

function _throwError([errorMsg, ...errorArgs]) {
    throw new Error( format(errorMsg || 'Assertion failed', ...errorArgs) );
}

export default {
    assert: function(expr, ...msgArgs) {
        expr || _throwError(msgArgs);
    },
    assertDefined: function(expr, ...msgArgs) {
        (typeof expr !== 'undefined') || _throwError(msgArgs);
    },
    assertBoolean: function(expr, ...msgArgs) {
        (typeof expr === 'boolean') || _throwError(msgArgs);
    },
    assertBooleanOrUndefined: function(expr, ...msgArgs) {
        (typeof expr === 'boolean' || typeof expr === 'undefined') || _throwError(msgArgs);
    },
    assertString: function(expr, ...msgArgs) {
        (typeof expr === 'string') || _throwError(msgArgs);
    },
    assertNonEmptyString: function(expr, ...msgArgs) {
        (expr && (typeof expr === 'string')) || _throwError(msgArgs);
    },
    assertStringOrEmpty: function(expr, ...msgArgs) {
        (!expr || typeof expr === 'string') || _throwError(msgArgs);
    },
    assertNumber: function(expr, ...msgArgs) {
        (typeof expr === 'number') || _throwError(msgArgs);
    },
    assertNumberInRange: function(expr, min, max, ...msgArgs) {
        (typeof expr === 'number' && expr >= min && expr <= max) || _throwError(msgArgs);
    },
    assertNumberInRangeOrEmpty: function(expr, min, max, ...msgArgs) {
        (!expr || (typeof expr === 'number' && expr >= min && expr <= max)) || _throwError(msgArgs);
    },
    assertFunction: function(expr, ...msgArgs) {
        (typeof expr === 'function') || _throwError(msgArgs);
    },
    assertFunctionOrEmpty: function(expr, ...msgArgs) {
        (!expr || (typeof expr === 'function')) || _throwError(msgArgs);
    },
    assertObject: function(expr, ...msgArgs) {
        (expr && typeof expr === 'object') || _throwError(msgArgs);
    },
    assertObjectOrEmpty: function(expr, ...msgArgs) {
        (!expr || (typeof expr === 'object')) || _throwError(msgArgs);
    },
    assertArray: function(expr, ...msgArgs) {
        (expr && Array.isArray(expr)) || _throwError(msgArgs);
    },
    assertArrayOrEmpty: function(expr, ...msgArgs) {
        (!expr || expr instanceof Array) || _throwError(msgArgs);
    },
    assertElement: function(expr, ...msgArgs) {
        (expr && expr.nodeType === 1) || _throwError(msgArgs);
    }
}
