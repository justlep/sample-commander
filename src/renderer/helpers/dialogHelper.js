import { IS_WINDOWS } from '@/constants'
const { dialog, getCurrentWindow } = require('electron').remote;

/**
 * @param {string} [title]
 * @param {string} [preselectedPath] 
 * @return {Promise<any>}
 */
export function selectSingleDirectory({title = 'Select directory', preselectedPath = undefined}) {
    return new Promise((resolve, reject) => {
        dialog.showOpenDialog(getCurrentWindow(), {
            title,
            defaultPath: preselectedPath || undefined,
            properties: ['openDirectory']
        }, dirs => {
            if (dirs && dirs.length) {
                resolve( dirs[0] );
            } else {
                reject('No directory selected');
            }
        });
    });
}

/**
 * @param {string} title 
 * @param {boolean} executable - if true, only executable file extensions will be selectable 
 * @return {Promise<any>}
 * TODO add file extension filter option
 *       filters: [
 *          {name: 'xml', extensions: ['xml'] }
 *       ]
 */
export function selectSingleFile({title = 'Select file', executable = false}) {
    let filters;

    if (executable && IS_WINDOWS) {
        filters = [{
            name: 'executables',
            extensions: ['exe', 'bat', 'cmd']
        }];
    }
    
    return new Promise((resolve, reject) => {
        dialog.showOpenDialog(getCurrentWindow(), {
            title,
            filters,
            properties: ['openFile']
        }, files => {
            if (files && files.length) {
                resolve( files[0] );
            } else {
                reject('No file selected');
            }
        });
    });
}
