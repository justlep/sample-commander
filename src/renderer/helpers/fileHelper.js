import path from 'path'
const promisify = require('util').promisify;
const fs = require('fs');

// const glob = require('glob');
// const readFile = promisify(fs.readFile);
// const writeFile = promisify(fs.writeFile);
export const copyFile = promisify(fs.copyFile);
export const mkdir = promisify(fs.mkdir);
export const unlink = promisify(fs.unlink);
export const stat = promisify(fs.stat);
export const rename = promisify(fs.rename);
export const access = promisify(fs.access);

export const COPYFILE_NO_OVERWRITE_FLAG = fs.constants.COPYFILE_EXCL;

export async function pathExists(targetPath) {
    if (!targetPath || typeof targetPath !== 'string') {
        throw 'Invalid path';
    }

    let fullPath = path.resolve(targetPath),
        exists = false;

    try {
        let statForPath = fullPath && await stat(fullPath);
        exists = !!(statForPath && (statForPath.isFile() || statForPath.isDirectory()));
    } catch (err) {
        // not ok
    }
    return exists;
}

/**
 * @param {string} path
 * @return {Promise<*>} resolves if executable, otherwise rejects
 */
export function assertFileIsExecutable(path) {
    if (path && typeof path === 'string') {
        return access(path, fs.constants.X_OK).then(() => path);
    }
    return Promise.reject();
}

/**
 * @param {string} targetPath
 * @param {boolean} expectFile
 * @return {Promise<string>} - fulfils with the validated full path
 * @private
 */
async function _assertPathExists(targetPath, expectFile) {
    if (!targetPath || typeof targetPath !== 'string') {
        throw 'Invalid path';
    }

    let fullPath = path.resolve(targetPath),
        isOk = false;

    try {
        let statForPath = fullPath && await stat(fullPath);
        isOk = statForPath && ((expectFile && statForPath.isFile()) || (!expectFile && statForPath.isDirectory()));
    } catch (err) {
        // not ok
    }
    if (!isOk) {
        throw 'Given path is not a ' + (expectFile ? 'file' : 'directory');
    }
    return fullPath;
}

/**
 * @param filePath
 * @return {Promise<string>} - the file path if exists, otherwise rejects
 */
export function assertFileExists(filePath) {
    return _assertPathExists(filePath, true);
}

/**
 * @param dirPath
 * @return {Promise<string>} - the dir path if exists, otherwise rejects
 */
export function assertDirExists(dirPath) {
    return _assertPathExists(dirPath, false);
}



