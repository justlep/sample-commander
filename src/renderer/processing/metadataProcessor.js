import {execFile} from 'child_process';
import Logger from '@/helpers/Logger'
import {format} from 'util';

/**
 * Determines metadata (e.g. sample rate) for this file using the ffprobe executable.
 * If the metadata is in the cache, that cached version is returned.
 *
 * @param {FileItem} fileItem
 * @param {string} ffprobeExecutablePath
 * @returns {Promise<{duration: String, sampleRate: Number}>}
 */
export async function getMetadataForFileItem(fileItem, ffprobeExecutablePath) {
    if (!ffprobeExecutablePath) {
        throw new Error('Invalid ffprobeExecutablePath');
    }

    let metadataJson;
    
    try {
        let metadataJsonString = await new Promise((resolve, reject) => {
                Logger.dev('Running ffprobe for metadata of %s', fileItem.path);
    
                execFile(ffprobeExecutablePath, [
                    '-hide_banner',
                    '-show_streams',
                    '-select_streams', 'a:0',
                    // '-sexagesimal',  // <-- for duration getting formatted like "0:09:39.396984" 
                    '-show_entries', 'stream=duration,duration_ts,sample_rate,time_base',
                    '-print_format', 'json',
                    fileItem.path
                ], {
                    timeout: 3000,
                    shell: false,
                    encoding: 'utf8',
                    stdio: ['pipe', 'pipe', 'ignore']
                }, (err, stdout, stderr) => {
                    if (err) {
                        reject(format('ffprobe failed to load metadata for %s, error: %s', fileItem.path, stderr));
                    } else {
                        resolve('' + stdout);
                    }
                });
            }),
            json = JSON.parse(metadataJsonString),
            firstStream = json && json.streams[0],
            sampleRate,
            duration,
            durationInMins;
        
        if (firstStream) {
            sampleRate = firstStream && parseInt(firstStream['sample_rate'], 10);
            
            let durationInSecs = parseFloat(firstStream.duration);

            // Logger.dev(firstStream);
            
            if (!isNaN(durationInSecs)) {
                // duration is in minutes
                let hours = (durationInSecs / 3600) << 0, 
                    mins = ((durationInSecs - (hours * 3600)) / 60) << 0, 
                    secs = (durationInSecs % 60) << 0,
                    parts = [];

                if (durationInSecs >= 1) {
                    if (hours) {
                        parts.push(hours + 'h');
                    }
                    if (mins) {
                        parts.push(mins + 'm');
                    }
                    if (secs) {
                        parts.push(secs + 's');
                    }
                } else {
                    parts.push(((durationInSecs * 1000) << 0) + 'ms');
                }
                duration = parts.join(' ');
                durationInMins = durationInSecs / 60;
            }
        }

        metadataJson = {
            sampleRate: sampleRate || null, 
            duration,
            durationInMins
        };
        
    } catch (err) {
        Logger.error('Failed to process metadata: %o', err);
        throw new Error(format('Failed to process metadata JSON  for %s, error: %s', fileItem.path, err));
    }
    
    // Logger.dev('Finished metadata for %s -> %o', fileItem.filename, metadataJson);
    return metadataJson;
}
