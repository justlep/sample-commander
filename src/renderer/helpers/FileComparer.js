import asserts from './asserts'
import nodeFs from 'fs'
import nodeCrypto from 'crypto'
import nodePath from 'path'

export const TOTAL_MIDDLE_HASH_BYTES = 10 * 1024; // must be round

export const HASH_ALGORITHM = 'sha256';
export const EMPTY_FILE_HASH = '_ZERO';

/** 
 * Map of fileItem.id -> middle-hash
 * @type {Object<string,string>} 
 **/
const middleHashCache = {}; // map of '{fileItem.id}' -> {middleHash}

/**
 * Map of fileItem.id -> full-hash
 * @type {Object<string,string>}
 **/
const fullHashCache = {};   // map of '{fileItem.id}' -> {fullHash}

/**
 * @param {number} fileSize
 * @param {number} totalBytesToHash
 * @returns {{startIndex: number, endIndex: number}} - the start index to read from, and end index (inclusive) 
 */
export function calculateReadIndexesForHashing(fileSize, totalBytesToHash) {
    if (totalBytesToHash % 2) {
        totalBytesToHash++;
    }
    if (!fileSize || !totalBytesToHash) {
        return {startIndex: 0, endIndex: 0};
    }
    let startIndex = Math.max(0, Math.min(Math.floor(fileSize / 2 - totalBytesToHash / 2), fileSize - totalBytesToHash)), 
        endIndex = Math.min(fileSize - 1, startIndex + totalBytesToHash - 1);
    
    return {startIndex, endIndex};
}

/**
 * Calculates the middle- or full hash for a given file item, then passes it to the callback.
 * Middle = hash of the center {@link TOTAL_MIDDLE_HASH_BYTES} bytes of the file
 * Full = hash over the entire file content
 * @param {FileItem} fileItem
 * @param {boolean} useFullHash - if true, the "full hash" is calculated instead of just the middle hash
 * @param {boolean} [noCache] - if true, the old cached value is replaced with a fresh, re-calculated one
 * @return {Promise<string>} - resolves with the calculated (or cached) hash 
 */
function calculateHash({fileItem, useFullHash, noCache}) {
    let filePath = nodePath.resolve(fileItem.path),
        fileSize = fileItem.filesize || 0,
        cacheKey = fileItem.id,
        cache = (useFullHash) ? fullHashCache : middleHashCache;

    if (!noCache && cache[cacheKey]) {
        return Promise.resolve(cache[cacheKey]);
    }

    if (!fileSize) {
        cache[cacheKey] = EMPTY_FILE_HASH;
        return Promise.resolve(cache[cacheKey]);
    }

    return new Promise((resolve, reject) => {
        let hash = nodeCrypto.createHash(HASH_ALGORITHM),
            {startIndex, endIndex} = calculateReadIndexesForHashing(fileSize, useFullHash ? fileSize : TOTAL_MIDDLE_HASH_BYTES),    
            readStream;
        
        try {
            readStream = nodeFs.ReadStream(filePath, {
                start: startIndex,
                end: endIndex 
            });
        } catch (err) {
            return reject(err);
        }
        
        // console.log('Reading offsets %s till %s of %s', streamOpts.start, streamOpts.end, fileItem.filesize);
        
        readStream.on('data', chunk => hash.update(chunk));
        readStream.on('error', err => reject(err));
        readStream.on('end', () => {
            cache[cacheKey] = hash.digest('hex');
            resolve(cache[cacheKey]);
        });
    });
}

/**
 * @param {FileItem} fileItem
 * @param {boolean} [noCache]
 * @return {Promise<string>}
 */
export function calculateMiddleHash({fileItem, noCache}) {
    return calculateHash({fileItem, useFullHash: false, noCache});
}

export function calculateFullHash({fileItem, noCache}) {
    return calculateHash({fileItem, useFullHash: true, noCache});
}

/**
 * Finds duplicate (target) file items for a given list of source FileItems, 
 * adding them to the source FileItems' duplicates.
 * 
 * @param {FileItem[]} sourceItems (Array) of source file items
 * @param {FileItem[]} targetItems (Array) of target file items
 * @param {CancellationToken} cancellationToken
 * @param {boolean} [quickMode=false] - if true, only the middle hash will be used for content comparison,
 *                                      otherwise full file content
 * @return {void} - returns when all files are compared
 */
export async function checkDuplicateFileItems({sourceItems, targetItems, cancellationToken, quickMode = false}) {
    asserts.assertArray(sourceItems, 'invalid sourceItems for checkDuplicateFileItems');
    asserts.assertArray(targetItems, 'invalid targetItems for checkDuplicateFileItems');

    const _hashingFn = quickMode ? calculateMiddleHash : calculateFullHash;

    // console.warn(`Starting file comparer in ${quickMode ? 'QUICK' : 'FULL CONTENT'} mode`);
    
    if (cancellationToken.isCancelled()) {
        return;
    }
    
    if (!sourceItems.length) {
        return;
    }

    // map of targetSize -> [targetItem, ..]
    let targetItemsBySize = targetItems.reduce((size2items, targetItem) => {
        let {filesize} = targetItem,
            sameSizeItems = size2items[filesize] || (size2items[filesize] = []);
        
        sameSizeItems.push(targetItem);
        return size2items;
    }, {}); 

    if (cancellationToken.isCancelled()) {
        return;
    }
    
    for (let sourceItem of sourceItems) {
        sourceItem.resetDuplicates();

        let sameSizeTargetItems = targetItemsBySize[sourceItem.filesize],
            sourceHash;
        
        if (!sameSizeTargetItems) {
            continue;
        }
        
        // TODO add try/catch

        for (let targetItem of sameSizeTargetItems) {
            if (sourceItem.path === targetItem.path) {
                // target file IS the source file 
                continue;
            }
            
            if (!sourceHash) {
                if (cancellationToken.isCancelled()) {
                    return;
                }
                sourceHash = await _hashingFn({fileItem: sourceItem});
            }

            if (cancellationToken.isCancelled()) {
                return;
            }
            
            let targetHash = await _hashingFn({fileItem: targetItem}),
                isIdentical = sourceHash === targetHash;

            if (isIdentical) {
                sourceItem.addDuplicate(targetItem);
            }
        }
    }
}

/**
 * @param {FileItem[]} fileItems
 */
export function flushCacheForFileItems(fileItems) {
    asserts.assertArray(fileItems, 'invalid fileItems for flushCacheForFileItems');
    for (let fileItem of fileItems) {
        delete middleHashCache[fileItem.id];
        delete fullHashCache[fileItem.id];
    }
}
