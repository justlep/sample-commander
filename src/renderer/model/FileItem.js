import nodeCrypto from 'crypto'
import nodePath from 'path'
import { assertDirExists, stat } from '@/helpers/fileHelper'
import readdirp from 'readdirp'
import {NON_REACTIVE_PROPS, AUDIO_FILE_EXTENSIONS_REGEX} from '@/constants'
import Logger from '@/helpers/Logger'
import asserts from '@/helpers/asserts'

/**
 * Represents non-reactive (audio) file items.
 */
export default class FileItem {
    /**
     * @param {string} audioFilePathOrName - an audio file path or filename
     * @return {string}
     */
    static getSpectrogramPathByAudioFilePath(audioFilePathOrName) {
        return audioFilePathOrName + '.SPX.png';
    }
    
    /**
     * @param {FileItem[]} fileItems
     * @return {FileItem[]}
     */
    static sortList(fileItems) {
        return fileItems.sort((a, b) => a.path.localeCompare(b.path));
    }

    /**
     * @param {string} path
     * @param {boolean} [recursive=true]
     * @param {RegExp} [filenamePattern] - defaults to {@link AUDIO_FILE_EXTENSIONS_REGEX}
     * @param {number} [limit] - the maximum number of files to load (default = Number.MAX_SAFE_INTEGER}, 0 means no limit
     * @param {CancellationToken} [cancellationToken]
     * @param {FileItem[]} [onlySameSizesAs] - an optional array of file items whose sizes are used to pre-filter
     *                                         the scanned files. I.e. if given, only files contained in the given FileItem array
     *                                         will be returned. (Intended primarily for preparing duplicate checks)
     * @return {Promise<{dirItems: FileItem[], isLimitExceeded: Boolean}>}
     */
    static findInPath({path, filenamePattern = AUDIO_FILE_EXTENSIONS_REGEX, recursive = true, 
                          limit = Number.MAX_SAFE_INTEGER, cancellationToken, onlySameSizesAs}) {
        if (isNaN(limit)) {
            return Promise.reject('Invalid limit ' + limit);
        }
        const effectiveLimit = limit || Number.MAX_SAFE_INTEGER;
        const allowedFileSizesMap = Array.isArray(onlySameSizesAs) ? onlySameSizesAs.reduce((sizeMap, fileItem) => {
            sizeMap[fileItem.filesize] = 1;
            return sizeMap;
        }, {}) : null;
        
        return assertDirExists(path).then(() => new Promise((resolve, reject) => {
                let fileItems = [],
                    isLimitExceeded = false,
                    stream = readdirp(path, {
                        type: 'files',
                        fileFilter: allowedFileSizesMap ?
                            entryInfo => allowedFileSizesMap[entryInfo.stats.size || 0] && filenamePattern.test(entryInfo.basename) :
                            entryInfo => filenamePattern.test(entryInfo.basename),
                        depth: recursive ? Number.MAX_SAFE_INTEGER : 0,
                        alwaysStat: true,
                        /** (!) .asar files break readdirp -> see comment in {@link DirItem#loadRecursively} */ 
                        directoryFilter: ['!*.asar'] 
                    }),
                    finishedCallback = () => {
                        FileItem.sortList(fileItems);
                        resolve({fileItems, isLimitExceeded});
                    },
                    /**
                     * @param {Object} entryInfo - https://github.com/paulmillr/readdirp#entryinfo
                     */
                    onData = entryInfo => {
                        isLimitExceeded = fileItems.length >= effectiveLimit;
                        if (isLimitExceeded || (cancellationToken && cancellationToken.isCancelled())) {
                            Logger.debug('destroy file stream (cancelled or limit exceeded)');
                            stream.pause();
                            stream.destroy();
                            return;
                        }

                        let fileItem = new FileItem({
                            path: entryInfo.fullPath,
                            size: entryInfo.stats.size || 0,
                            mtime: entryInfo.stats.mtime.getTime(),
                            validate: !fileItems.length // validate first item to notice future breaking changes in readdirp
                        });

                        fileItems.push(fileItem);
                    };

            stream
                .on('data', onData)
                .on('warn', err => Logger.warn(err))
                .on('error', err => reject(err)) 
                //.on('end', finishedCallback)
                .on('close', () => {
                    Logger.debug('close FileItem stream');
                    finishedCallback();
                })
                .on('end', () => {
                    Logger.debug('end FileItem stream');
                    finishedCallback();
                });
            })
        );
    }

    /**
     * @param {string[]} paths
     * @returns {Promise<FileItem[]>}
     */
    static async getByPaths({paths}) {
        let fileItems = [];
        for (let filePath of paths) {
            try {
                let fileStat = await stat(filePath);
                if (fileStat.isFile()) {
                    fileItems.push(new FileItem({
                        path: filePath,
                        size: fileStat.size,
                        mtime: fileStat.mtime
                    }));
                }
            } catch (err) {
                Logger.dev('Ignoring missing file: %s', filePath);
            }
        }
        return fileItems;
    }
    
    /**
     * @param {string} path
     * @param {number} size
     * @param {number} mtime
     * @param {boolean} [validate] - set true to validate parameters and throw an error if invalid 
     */
    constructor({path, size = 0, mtime, validate}) {
        let resolvedPath = nodePath.resolve(path),
            idHash = nodeCrypto.createHash('sha1'),
            isAudioFile = AUDIO_FILE_EXTENSIONS_REGEX.test(resolvedPath),
            supportsSpectrograms = isAudioFile;

        if (validate) {
            asserts.assertNonEmptyString(path);
            asserts.assertNumberInRange(size, 0, Number.MAX_SAFE_INTEGER);
            asserts.assertNumberInRange(mtime, 0, Number.MAX_SAFE_INTEGER);
            // Logger.debug(path, size, mtime);
        }
        
        idHash.update(resolvedPath + '::' + size + '::' + mtime + '::');
        
        this[NON_REACTIVE_PROPS] = {
            id: idHash.digest('hex'),
            filename: nodePath.basename(resolvedPath),
            path: resolvedPath,
            parentDir: nodePath.join(resolvedPath, '..'),
            filesize: size,
            formattedMTime: new Date(mtime).toLocaleString(), // moment(mtime).format('DD.MM.YYYY - hh:mm:ss (ddd)'),
            formattedFilesize: (size / (1024 * 1024)).toFixed(1),
            isAudioFile,
            supportsSpectrograms,
            isDeleted: false,
            renamedIntoPath: null
        };
        
        this.metadata = {
            duration: '',
            sampleRate: ''
        };
        this.spectrogram = null;
        
        /** @type {String[]} */
        this._duplicatePaths = null;
    }
    
    get id() { return this[NON_REACTIVE_PROPS].id }
    get isAudioFile() { return this[NON_REACTIVE_PROPS].isAudioFile }
    get filename() { return this[NON_REACTIVE_PROPS].filename }
    get path() { return this[NON_REACTIVE_PROPS].path }
    get parentDir() { return this[NON_REACTIVE_PROPS].parentDir }
    get filesize() { return this[NON_REACTIVE_PROPS].filesize }
    get formattedMTime() { return this[NON_REACTIVE_PROPS].formattedMTime }
    get formattedFilesize() { return this[NON_REACTIVE_PROPS].formattedFilesize }
    get supportsSpectrograms() { return this[NON_REACTIVE_PROPS].supportsSpectrograms }
    
    get isDeleted() { return this[NON_REACTIVE_PROPS].isDeleted }
    get renamedIntoPath() { return this[NON_REACTIVE_PROPS].renamedIntoPath }

    hasMetadata() {
        return !!(this.metadata.duration || this.metadata.sampleRate);
    }

    /**
     * @returns {number|null} - the file's Nyquist frequency or null if the `metadata.sampleRate` isn't set 
     */
    getNyquistFrequency() {
        let sampleRate = this.metadata.sampleRate;
        return isNaN(sampleRate) ? null : sampleRate / 2;
    }
    
    /**
     * Returns all registered duplicate paths, but excluding its own path.
     * @return {String[]}
     */
    getDuplicatePathsWithoutSelf() {
        return (this._duplicatePaths || []).filter(path => path !== this.path);
    }
    
    /**
     * @param {FileItem} fileItem
     */
    addDuplicate(fileItem) {
        if (this._duplicatePaths) {
            this._duplicatePaths.push(fileItem.path);
        } else {
            this._duplicatePaths = [fileItem.path];
        }
    }
    
    resetDuplicates() {
        this._duplicatePaths = null;
    }
    
    /** @type {boolean} */
    hasDuplicates() {
        return !!(this._duplicatePaths && this._duplicatePaths.length);
    }
    
    /** @type {boolean} */
    isUnique() {
        return !this._duplicatePaths || !this._duplicatePaths.length;
    }

    /**
     * @return {string}
     */
    getSpectrogramFilename() {
        return FileItem.getSpectrogramPathByAudioFilePath(this.filename);
    }
    
    /**
     * @return {string}
     */
    getSpectrogramPath() {
        return FileItem.getSpectrogramPathByAudioFilePath(this.path);
    }

    setDeleted() {
        this[NON_REACTIVE_PROPS].isDeleted = true;
    }

    /**
     * Marks this file item to be renamed, so it can be cleaned up and re-looked-up.
     * @param {string} newPath
     */
    setRenamed(newPath) {
        this[NON_REACTIVE_PROPS].renamedIntoPath = newPath;
    }
    
    /**
     * @override
     */
    toString() {
        return 'FileItem['+ this.id +']';
    }
}


/**
 * @typedef {Object} Readdirp_EntryInfo
 * @property {string} parentDir - directory in which entry was found (relative to given root)
 * @property {string} fullParentDir - full path to parent directory
 * @property {string} name - name of the file/directory
 * @property {string} path - path to the file/directory (relative to given root)
 * @property {string} fullPath - full path to the file/directory found
 * @property {Object} stat - built in stat object
 */
