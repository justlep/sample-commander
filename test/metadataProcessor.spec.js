import {getMetadataForFileItem} from '@/processing/metadataProcessor'
import FileItem from '@/model/FileItem'
import {assertFileIsExecutable} from '@/helpers/fileHelper'
import path from 'path'

const TEST_FILES_DIR_PATH = path.join(__dirname, 'test-files');

const FFPROBE_EXECUTABLE_PATH = path.resolve('c:\\progs\\ffmpeg-4.1.1-win64-static\\bin\\ffprobe.exe');

describe('metadataProcessor', () => {
   
    xit('can determine metadata for audio fileItems using ffprobe.exe', async () => {
        await assertFileIsExecutable(FFPROBE_EXECUTABLE_PATH);
        
        let {fileItems, isLimitExceeded} = await FileItem.findInPath({path: TEST_FILES_DIR_PATH});
        expect(fileItems.length).toBeGreaterThan(0);
        
        let allMetadata = [];
        
        for (let fileItem of fileItems) {
            expect(fileItem.hasMetadata()).toBe(false);
            let metadata = await getMetadataForFileItem(fileItem, FFPROBE_EXECUTABLE_PATH);
            
            fileItem.metadata = metadata;
            expect(fileItem.hasMetadata()).toBe(true);
            
            allMetadata.push({
                filename: fileItem.filename,
                size: fileItem.filesize,
                nyquistFrequency: fileItem.getNyquistFrequency(),
                metadata
            });
        }
        
        expect(allMetadata).toMatchSnapshot();
    });
    
});
