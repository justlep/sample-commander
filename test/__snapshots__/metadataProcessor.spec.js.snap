// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadataProcessor can determine metadata for audio fileItems using ffprobe.exe 1`] = `
Array [
  Object {
    "filename": "20k-sweep-11k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334083333333332,
      "sampleRate": 11025,
    },
    "nyquistFrequency": 5512.5,
    "size": 16584,
  },
  Object {
    "filename": "20k-sweep-192k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334116666666667,
      "sampleRate": 192000,
    },
    "nyquistFrequency": 96000,
    "size": 288072,
  },
  Object {
    "filename": "20k-sweep-22k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334083333333332,
      "sampleRate": 22050,
    },
    "nyquistFrequency": 11025,
    "size": 33122,
  },
  Object {
    "filename": "20k-sweep-44k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334083333333332,
      "sampleRate": 44100,
    },
    "nyquistFrequency": 22050,
    "size": 66200,
  },
  Object {
    "filename": "20k-sweep-48k-copy.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334033333333333,
      "sampleRate": 48000,
    },
    "nyquistFrequency": 24000,
    "size": 72050,
  },
  Object {
    "filename": "20k-sweep-48k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334033333333333,
      "sampleRate": 48000,
    },
    "nyquistFrequency": 24000,
    "size": 72050,
  },
  Object {
    "filename": "20k-sweep-96k.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334033333333333,
      "sampleRate": 96000,
    },
    "nyquistFrequency": 48000,
    "size": 144056,
  },
  Object {
    "filename": "20k-sweep-48k-in-subdir.mp3",
    "metadata": Object {
      "duration": "528ms",
      "durationInMins": 0.0088,
      "sampleRate": 48000,
    },
    "nyquistFrequency": 24000,
    "size": 4461,
  },
  Object {
    "filename": "20k-sweep-48k-in-subdir.wav",
    "metadata": Object {
      "duration": "500ms",
      "durationInMins": 0.008334033333333333,
      "sampleRate": 48000,
    },
    "nyquistFrequency": 24000,
    "size": 72050,
  },
]
`;
